### ISSUE TITLE
Question: Type signature for path-like objects?

### ISSUE THREAD

**Issue by madig** [microsoft/pyright#1298]
Pylance rejects the following piece of code:

```python
import os
from typing import Union

def foo(path: Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]) -> None:
    os.fspath(path)
```

```
No overloads for "os.fspath(path)" match parameters
  Argument types: (str | bytes | _PathLike[str] | _PathLike[bytes])
```

Looking into typeshed (`os/__init__.pyi` and `builtins.pyi`) makes me think that it _should_ work:

```python
@overload
def fspath(path: str) -> str: ...
@overload
def fspath(path: bytes) -> bytes: ...
@overload
def fspath(path: PathLike[AnyStr]) -> AnyStr: ...
```

```python
# Predefined type variables.
AnyStr = TypeVar("AnyStr", str, bytes)

# [...] 

# This class is to be exported as PathLike from os,
# but we define it here as _PathLike to avoid import cycle issues.
# See https://github.com/python/typeshed/pull/991#issuecomment-288160993
_AnyStr_co = TypeVar("_AnyStr_co", str, bytes, covariant=True)
@runtime_checkable
class _PathLike(Protocol[_AnyStr_co]):
    def __fspath__(self) -> _AnyStr_co: ...
```

I asked in the `python/typing` Gitter channel and was told that my signature looks fine. What am I missing?

**Comment by erictraut** [comment]
Pyright's handling of overloads differs subtly from that of mypy, and you're seeing the effect of that difference. Pyright selects a single overload based on the argument types. Mypy will apply multiple overloads for a single call if some of the argument types are unions. The "correct" behavior isn't specified in PEP 484, so it's left up to the type checker author to decide how to implement overload selection. Pyright's current behavior matches the handling of overloads in other languages like TypeScript. I've thought about changing pyright to match mypy's behavior for consistency, but it's a very large change, and I have concerns about the performance implications because of the potential for combinatoric explosions when presented with many arguments that all have union types with many subtypes.

**Comment by madig** [comment]
So essentially, typeshed has a mypy-bias. Hm. Is there a way to satisfy Pyright somehow in this instance?

**Comment by erictraut** [comment]
I can't think of any elegant workarounds. Options include a call to `cast` and a series of `if isinstance` checks that narrow the type.

Incidentally, while I've known about this issue for a long time, this is only the second time someone has asked about it in two years. Plus, we have never hit this issue in my team's half-a-million-line Python source base, so I don't think it's an issue that arises frequently.

**Comment by madig** [comment]
Oh well. Thanks for looking into it.

**Comment by erictraut** [comment]
After receiving a number of bug reports on this issue, I've decided to change the behavior of overload matching to more closely mirror that of other type checkers. New versions of of pyright will perform union expansion and evaluate multiple overloads. I've addressed the combinatoric explosion problem by limiting expansion to 64 combinations. In practice, this will occur very infrequently because it would require more than 6 arguments with two union subtypes each or more than 2 arguments with 8 union subtypes each.
